### ISSUE TITLE
error: expected: !is_cycle_free()

### ISSUE THREAD

**Issue by peter-ahe-google** [dart-lang/kernel#38]
```
$ dart --version
Dart VM version: 1.20.0-edge.f922a22fdbffd81c8a781df4cb8468223fb1ca96 (Wed Oct  5 15:09:41 2016) on "macos_x64"
$ dart works.dill 
$ dart crash.dill 
.../sdk/runtime/vm/object.cc: 3601: error: expected: !is_cycle_free()
Dumping native stack trace for thread 50f
  [0x000000010e7f123d] dart::Profiler::DumpStackTrace(bool)
  [0x000000010e7f123d] dart::Profiler::DumpStackTrace(bool)
-- End of DumpStackTrace
Abort trap: 6
```

[repro.zip](https://github.com/dart-lang/kernel/files/513315/repro.zip)

The key difference between these two files is whether or not I've applied closure conversion to `dart:core::Object::_symbolMapToStringMap`:

```
--- works.txt   2016-10-06 11:27:15.000000000 +0200
+++ crash.txt   2016-10-06 11:27:21.000000000 +0200
@@ -7083,11 +7083,12 @@
     @_in::ExternalName::•("Object_as")
     external method _as(dynamic instantiator_type_arguments, dynamic type) → dynamic;
     static method _symbolMapToStringMap(core::Map<core::Symbol, dynamic> map) → dynamic {
-      dynamic result = core::Map::•<core::String, dynamic>();
-      map.forEach((core::Symbol key, dynamic value) → dynamic {
-        result.[]=(_in::Symbol::getName(key), value);
-      });
-      return result;
+      final mock::Context #t107 = new mock::Context::•(1);
+      #t107.parent = null;
+      #t107.[]=(0, core::Map::•<core::String, dynamic>());
+      dynamic #t108 = invalid-expression;
+      map.forEach(new core::Closure#2::•(#t107));
+      return #t107.[](0);
     }
   }
   abstract class Pattern extends core::Object { // from file:///Users/ahe/kernel-sdk/sdk/xcodebuild/DerivedSources/DebugX64/patched_sdk/lib/core/pattern.dart
```

**Comment by peter-ahe-google** [comment]
Reproduces at https://github.com/dart-lang/kernel_sdk/commit/f922a22fdbffd81c8a781df4cb8468223fb1ca96.

**Comment by peter-ahe-google** [comment]
Fix reverted in f16ba58.

**Comment by kmillikin** [comment]
Fixed by https://github.com/dart-lang/kernel_sdk/commit/0cacbaaf228a75bd2c9b0a4a06eac1227dceca83
